text/3660-crates-io-crate-deletions.md

```diff
@@ -0,0 +1,123 @@
+
+This restriction makes sense for the majority of crates that have been around for a while and are actively used, but the above list of reasons shows that there are valid use cases for allowing crate authors to delete their crates without having to contact the crates.io team.
+
+To make this process easier for our users and to reduce the workload of the crates.io team dealing with such support requests, we propose to codify our current set of informal rules into a formal policy that allows crate authors to delete their crates themselves under certain conditions (see below).
+
+
+# Proposal
+
+We propose to allow crate authors to delete their **crates** from crates.io under the following conditions:
+
+* The crate has been published for less than 72 hours,
+* or if all the following conditions are met:
+  * The crate has a single owner,
+  * The crate has no reverse dependencies,
+  * The crate has been downloaded less than 100 times for each month it has been published.
+
+We also propose to allow crate authors to delete **versions** of their crates from crates.io under the following conditions:
+
+* The version has been published for less than 72 hours.
+
+These crate owner actions will be enabled by two new API endpoints:
+
+- `DELETE /api/v1/crates/:crate_id` to delete a crate
+- `DELETE /api/v1/crates/:crate_id/:version` to delete a version
+
+
+# Drawbacks
+
+> Why should we *not* do this?
+
+The main drawback of this proposal is that it makes the crates.io registry less immutable.
+This could lead to confusion if a crate is deleted that is depended on by other projects that are not published on crates.io themselves.
+However, we believe that the conditions we propose are strict enough to prevent this from happening in practice due to the additional download threshold.
+
+Another potential drawback is that it can create confusion on when it would be better to yank a version instead of deleting it.
+We plan to address this by adding a note to the usage policy that explains the difference between yanking and deleting a version, and when to use which action based on the list in the [Motivation](#motivation) section above.
+
+
+# Rationale and alternatives
+
+> Why is this design the best in the space of possible designs?
+
+The proposed design is based on the current informal rules that the crates.io team uses to decide whether to delete a crate or version.
+These rules have been derived from the npm registry, which has a similar policy (see below).
+We believe that the proposed conditions are strict enough to prevent accidental deletions while still allowing crate authors to delete their crates in the cases where it makes sense.
+
+> What other designs have been considered and what is the rationale for not choosing them?
+
+We considered not having restrictions on the number of reverse dependencies, but since that would leave the package index in an inconsistent state, we decided to require that the crate has no reverse dependencies.
```

Comment:
The RFC as written does not seem to require an absence of reverse dependencies in all cases. In particular, the requirements for the new crate case seem to allow deletion if it's been less than 72 hours without any restriction about reverse dependencies, and the requirements for the crate-version case don't place any requirement on reverse dependencies either. (I posted a suggestion for the latter.)

Reply:
This still seems to be outdated or misleading? There's no requirement on reverse deps within 72 hours for either versions or new crates, both of which will likely lead to inconsistencies in the index.

Reply:
The RFC seems to have been merged without this being fixed. :(

Reply:
@Turbo87 It would have been nice to get some amount of feedback on the discussion that took place, rather than just seeing this merged with several threads completely ignored. From what I can tell, the tradeoff here is between this causing problems on the one side, and this requiring some additional server-side checks on the other side. Is there an actual user use case that is supported by deleting crates within 72 hours that have reverse dependencies? Or is this an implementation-bandwidth concern that could potentially be helped with by contributing an implementation of the restrictions on reverse dependencies?

Reply:
> It would have been nice to get some amount of feedback on the discussion that took place, rather than just seeing this merged with several threads completely ignored.

yeah, you are not the first one to point this out, and I agree. this was only the second RFC I've ever opened and I'm still learning how to handle them well. I'll try to do better in the future :)

regarding your concerns about version deletions: I've put this on the agenda again for the crates.io team meeting today.

> is this an implementation-bandwidth concern

no, while our bandwidth is limited, I don't see it as an issue in this particular case.

Reply:
> regarding your concerns about version deletions: I've put this on the agenda again for the crates.io team meeting today.

the crates.io team just discussed the topic of version deletions again and we concluded to entirely remove that part of the RFC. I'll open a PR to adjust the RFC accordingly. It might have to wait until after the weekend, but I wanted to give a heads-up here that this is coming. I'll go into more detail in that PR then.

Reply:
@Turbo87 Thank you very much! I think the policy for deleting *whole crates*, with the careful restrictions on them to make sure nobody is using them yet, seem completely fine; the concern was exclusively with deleting crate *versions* that might have dependencies on them.